File: cpp/section03/exercises/exercise03.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

void retrieveOldestAndNewestTweet(const vector<string>& array) {
    // O(1)
    cout << array[0] << " , " << array[array.size() - 1] << endl; 
}

struct Tweet {
    string tweet;
    int date;
};

void retrieveOldestAndNewestTweetWithDateComparison(const vector<Tweet>& array) {
    int newestDate = array[array.size() - 1].date; // Initialize with the first date
    int oldestDate = array[0].date; // Initialize with the first date

    for (const auto& item : array) { // O(n)
        if (item.date >= newestDate) {
            newestDate = item.date;
        }

        if (item.date <= oldestDate) {
            oldestDate = item.date;
        }
    }

    cout << "oldest date: " << oldestDate << " , newest date: " << newestDate << endl;
}

int main() {
    vector<string> tweets = {"hey", "wow", "hi", "nice"};
    retrieveOldestAndNewestTweet(tweets);

    vector<Tweet> tweetsWithDate = {
        {"hey", 2014},
        {"wow", 2012},
        {"hi", 2018}
    };

    retrieveOldestAndNewestTweetWithDateComparison(tweetsWithDate); // Time complexity = O(n) and space complexity = O(1)

    // Discussing the length property in C++
    string exampleString = "asdjfajlksdjlkcjalksdjkljjlk";
    cout << "Length of the string: " << exampleString.length() << endl; // O(1)

    /*
    The time complexity for accessing .length in C++ depends on how the string class is implemented.
    In C++, string::length() is typically O(1) because it returns a stored size attribute of the string.
    */
    
    return 0;
}

File: javascript/section03/exercises/exercise03.js
let tweets = ["hey", 'wow', 'hi', 'nice'];

const retrieveOldestAndNewestTweet = (array) => {
    console.log(array[0], ' , ', array[array.length - 1]); //o(1)
}

//retrieveOldestAndNewestTweet(tweets);


let tweetsWithDate = [{
    tweet: 'hey',
    date: 2014
    }, {
    tweet: 'Wow',
    date: 2012
    }, {
    tweet: 'hi',
    date: 2018
}];

const retrieveOldestAndNewestTweetWithDateComparison = (array) => { 
    let newestDate = array[array.length - 1].date; 
    let oldestDate = array[0].date; 
    array.forEach(item => { 
        if(item.date >= newestDate){
            newestDate = item.date;
        }

        if(item.date < oldestDate){
            oldestDate = item.date;
        }

    });
    console.log('oldest date', oldestDate, ' newest date: ', newestDate);
}

retrieveOldestAndNewestTweetWithDateComparison(tweetsWithDate);// time complexity = O(n) and space complexity = O(1)




// You may get asked, what is the big O of using the .length in this case:
'asdjfajlksdjlkcjalksdjkljjlk'.length; 
/*
The answer to this question is that it depends

It depends on the  language you are using 

We need to know how the method works on the string here and how it works in javascript or any other language it might use


the .length here in js is a property if you look in the docs

So it's only going there and retrieving 

So O(1) constant time
*/ 